File: types/settings/billing.ts
// Billing Types
// Sprint 7: Billing and Usage

// Plan tier identifiers
export type PlanTier = 'free' | 'pro' | 'teams' | 'enterprise';

// Plan billing interval
export type BillingInterval = 'monthly' | 'annually';

// Plan definition
export interface Plan {
  id: PlanTier;
  name: string;
  price: number;
  annualPrice?: number;
  description: string;
  features: string[];
  limits: {
    apiCalls: number;
    ingestions: number;
    storageGb: number;
    seats: number;
    oracleQueries: number;
  };
  isPopular?: boolean;
}

// Plan definitions with limits
export const PLANS: Plan[] = [
  {
    id: 'free',
    name: 'Free',
    price: 0,
    description: 'For individuals getting started',
    features: [
      'Basic context management',
      'Limited API calls',
      '1 GB storage',
      '1 seat',
      '10 Oracle queries/day',
    ],
    limits: {
      apiCalls: 1000,
      ingestions: 50,
      storageGb: 1,
      seats: 1,
      oracleQueries: 10,
    },
  },
  {
    id: 'pro',
    name: 'Pro',
    price: 29,
    annualPrice: 24,
    description: 'For professionals and power users',
    features: [
      'Advanced context management',
      '10,000 API calls/month',
      '25 GB storage',
      '3 seats',
      '100 Oracle queries/day',
      'Priority support',
    ],
    limits: {
      apiCalls: 10000,
      ingestions: 500,
      storageGb: 25,
      seats: 3,
      oracleQueries: 100,
    },
    isPopular: true,
  },
  {
    id: 'teams',
    name: 'Teams',
    price: 79,
    annualPrice: 66,
    description: 'For growing teams',
    features: [
      'Team collaboration',
      '50,000 API calls/month',
      '100 GB storage',
      '10 seats',
      'Unlimited Oracle queries',
      'SSO integration',
      'Advanced analytics',
    ],
    limits: {
      apiCalls: 50000,
      ingestions: 2500,
      storageGb: 100,
      seats: 10,
      oracleQueries: -1,
    },
  },
  {
    id: 'enterprise',
    name: 'Enterprise',
    price: -1,
    description: 'For large organizations',
    features: [
      'Custom limits',
      'Dedicated support',
      'SLA guarantees',
      'Custom integrations',
      'On-premise deployment',
      'Audit logs',
    ],
    limits: {
      apiCalls: -1,
      ingestions: -1,
      storageGb: -1,
      seats: -1,
      oracleQueries: -1,
    },
  },
];

// Subscription status
export type SubscriptionStatus = 'active' | 'canceled' | 'past_due' | 'trialing' | 'incomplete';

// Subscription entity
export interface Subscription {
  id: string;
  userId: string;
  planId: PlanTier;
  status: SubscriptionStatus;
  billingInterval: BillingInterval;
  stripeSubscriptionId?: string;
  currentPeriodStart: Date;
  currentPeriodEnd: Date;
  cancelAtPeriodEnd: boolean;
  trialEnd?: Date;
  createdAt: Date;
  updatedAt: Date;
}

// Payment method types
export type PaymentMethodType = 'card' | 'sepa_debit';

// Card brand
export type CardBrand = 'visa' | 'mastercard' | 'amex' | 'discover' | 'unknown';

// Payment method entity
export interface PaymentMethod {
  id: string;
  userId: string;
  type: PaymentMethodType;
  stripePaymentMethodId: string;
  isDefault: boolean;
  cardBrand?: CardBrand;
  cardLast4?: string;
  cardExpMonth?: number;
  cardExpYear?: number;
  sepaBankCode?: string;
  sepaLast4?: string;
  sepaCountry?: string;
  createdAt: Date;
}

// Invoice status
export type InvoiceStatus = 'draft' | 'open' | 'paid' | 'void' | 'uncollectible';

// Invoice entity
export interface Invoice {
  id: string;
  userId: string;
  stripeInvoiceId: string;
  invoiceNumber: string;
  status: InvoiceStatus;
  amountDue: number;
  amountPaid: number;
  currency: string;
  description?: string;
  invoiceUrl?: string;
  pdfUrl?: string;
  periodStart: Date;
  periodEnd: Date;
  dueDate?: Date;
  paidAt?: Date;
  createdAt: Date;
}

// Usage metric types
export type UsageMetricType = 'api_calls' | 'ingestions' | 'storage' | 'seats' | 'oracle_queries';

// Usage metric labels
export const USAGE_METRIC_LABELS: Record<UsageMetricType, { label: string; unit: string }> = {
  api_calls: { label: 'API Calls', unit: 'calls' },
  ingestions: { label: 'Ingestions', unit: 'imports' },
  storage: { label: 'Storage', unit: 'GB' },
  seats: { label: 'Seats', unit: 'members' },
  oracle_queries: { label: 'Oracle Queries', unit: 'queries/day' },
};

// Usage metric data point
export interface UsageDataPoint {
  date: Date;
  value: number;
}

// Usage metric
export interface UsageMetric {
  type: UsageMetricType;
  current: number;
  limit: number;
  percentage: number;
  trend: UsageDataPoint[];
}

// Billing details (Dutch BV compliance)
export interface BillingDetails {
  userId: string;
  companyName?: string;
  btwNumber?: string;
  kvkNumber?: string;
  addressLine1?: string;
  addressLine2?: string;
  city?: string;
  postalCode?: string;
  country: string;
  updatedAt: Date;
}

// BTW validation result
export interface BtwValidationResult {
  isValid: boolean;
  companyName?: string;
  address?: string;
  errorMessage?: string;
}

// Billing state for context
export interface BillingState {
  subscription: Subscription | null;
  paymentMethods: PaymentMethod[];
  invoices: Invoice[];
  usageMetrics: UsageMetric[];
  billingDetails: BillingDetails | null;
  isLoading: boolean;
  error: string | null;
}

// Billing actions
export type BillingAction =
  | { type: 'SET_SUBSCRIPTION'; payload: Subscription }
  | { type: 'SET_PAYMENT_METHODS'; payload: PaymentMethod[] }
  | { type: 'ADD_PAYMENT_METHOD'; payload: PaymentMethod }
  | { type: 'REMOVE_PAYMENT_METHOD'; payload: string }
  | { type: 'SET_DEFAULT_PAYMENT_METHOD'; payload: string }
  | { type: 'SET_INVOICES'; payload: Invoice[] }
  | { type: 'SET_USAGE_METRICS'; payload: UsageMetric[] }
  | { type: 'SET_BILLING_DETAILS'; payload: BillingDetails }
  | { type: 'SET_LOADING'; payload: boolean }
  | { type: 'SET_ERROR'; payload: string | null };

// Component Props for Billing Section

export interface CurrentPlanProps {
  subscription: Subscription | null;
  onUpgrade: (planId: PlanTier) => Promise<void>;
  onDowngrade: (planId: PlanTier) => Promise<void>;
  onCancelSubscription: () => Promise<void>;
  onResumeSubscription: () => Promise<void>;
}

export interface PlanComparisonModalProps {
  isOpen: boolean;
  currentPlanId: PlanTier;
  onClose: () => void;
  onSelectPlan: (planId: PlanTier) => Promise<void>;
}

export interface PaymentMethodsProps {
  paymentMethods: PaymentMethod[];
  onAddPaymentMethod: (type: PaymentMethodType) => Promise<void>;
  onRemovePaymentMethod: (id: string) => Promise<void>;
  onSetDefault: (id: string) => Promise<void>;
}

export interface AddPaymentMethodModalProps {
  isOpen: boolean;
  type: PaymentMethodType;
  onClose: () => void;
  onSubmit: () => Promise<void>;
}

export interface InvoiceHistoryProps {
  invoices: Invoice[];
  isLoading: boolean;
  onDownload: (invoiceId: string) => Promise<void>;
  onLoadMore: () => Promise<void>;
  hasMore: boolean;
}

export interface UsageMetricsProps {
  metrics: UsageMetric[];
  subscription: Subscription | null;
}

export interface UsageProgressBarProps {
  metric: UsageMetric;
  showTrend?: boolean;
}

export interface BillingDetailsProps {
  details: BillingDetails | null;
  onSave: (details: Partial<BillingDetails>) => Promise<void>;
  onValidateBtw: (btwNumber: string) => Promise<BtwValidationResult>;
}

File: server/app/models/billing.py
"""Billing domain Pydantic models matching TypeScript types/settings/billing.ts."""

from datetime import datetime
from decimal import Decimal
from typing import Any
from uuid import UUID

from pydantic import Field

from server.app.models.base import BaseSchema
from server.app.models.enums import (
    BillingInterval,
    PaymentMethodType,
    PlanTier,
    SubscriptionStatus,
)


# --- Plan ---

class PlanLimits(BaseSchema):
    """Plan tier limits matching TypeScript Plan.limits."""

    api_calls: int = Field(alias="apiCalls")
    ingestions: int
    storage_gb: int = Field(alias="storageGb")
    seats: int
    oracle_queries: int = Field(alias="oracleQueries")


class Plan(BaseSchema):
    """Plan definition matching TypeScript Plan interface."""

    id: PlanTier
    name: str
    price: Decimal  # Use Decimal, never float
    annual_price: Decimal | None = Field(None, alias="annualPrice")
    description: str
    features: list[str]
    limits: PlanLimits
    is_popular: bool | None = Field(None, alias="isPopular")


# --- Subscription ---

class Subscription(BaseSchema):
    """Current subscription details."""

    id: str  # Stripe subscription ID
    plan: PlanTier
    status: SubscriptionStatus
    billing_interval: BillingInterval
    current_period_start: datetime
    current_period_end: datetime
    cancel_at_period_end: bool
    trial_end: datetime | None = None


class SubscriptionCreateRequest(BaseSchema):
    """Request body for creating a checkout session."""

    plan: PlanTier
    billing_interval: BillingInterval = BillingInterval.MONTHLY
    success_url: str
    cancel_url: str


class SubscriptionUpdateRequest(BaseSchema):
    """Request body for updating subscription."""

    plan: PlanTier | None = None
    billing_interval: BillingInterval | None = None


# --- Payment Method ---

class PaymentMethod(BaseSchema):
    """Payment method details."""

    id: str  # Stripe payment method ID
    type: PaymentMethodType
    is_default: bool
    card_last4: str | None = None
    card_brand: str | None = None
    card_exp_month: int | None = None
    card_exp_year: int | None = None
    ideal_bank: str | None = None
    sepa_last4: str | None = None


class PaymentMethodCreateRequest(BaseSchema):
    """Request body for adding a payment method."""

    payment_method_id: str  # From Stripe.js


# --- Invoice ---

class Invoice(BaseSchema):
    """Invoice details."""

    id: str  # Stripe invoice ID
    number: str | None
    status: str
    amount_due: Decimal  # In cents, Decimal not float
    amount_paid: Decimal
    currency: str = "eur"
    period_start: datetime
    period_end: datetime
    hosted_invoice_url: str | None = None
    invoice_pdf: str | None = None
    created_at: datetime


# --- Usage ---

class UsageMetric(BaseSchema):
    """Current usage metrics per category."""

    category: str
    used: int
    limit: int
    percentage: float
    unit: str


class UsageOverview(BaseSchema):
    """Complete usage overview."""

    plan: PlanTier
    metrics: list[UsageMetric]
    billing_period_start: datetime
    billing_period_end: datetime


# --- Billing Details ---

class BillingDetails(BaseSchema):
    """Billing information for invoicing (Dutch BV compliance)."""

    company_name: str | None = None
    btw_number: str | None = None  # Dutch VAT number
    kvk_number: str | None = None  # Dutch Chamber of Commerce number
    address_line1: str | None = None
    address_line2: str | None = None
    city: str | None = None
    postal_code: str | None = None
    country: str = "NL"
    email: str | None = None


class BillingDetailsUpdate(BaseSchema):
    """Request body for updating billing details."""

    company_name: str | None = None
    btw_number: str | None = None
    kvk_number: str | None = None
    address_line1: str | None = None
    address_line2: str | None = None
    city: str | None = None
    postal_code: str | None = None
    country: str | None = None
    email: str | None = None


class BtwValidationResult(BaseSchema):
    """BTW (VAT) number validation result via VIES API."""

    valid: bool
    btw_number: str
    company_name: str | None = None
    company_address: str | None = None
    message: str | None = None


# --- Checkout Session ---

class CheckoutSessionResponse(BaseSchema):
    """Stripe checkout session response."""

    session_id: str
    url: str
